### PR TITLE
[SPARK-9301] [SQL] Add collect_set and collect_list aggregate functions

### PR DESCRIPTION
**Pull request by nburoojy**
For now they are thin wrappers around the corresponding Hive UDAFs.

One limitation with these in Hive 0.13.0 is they only support aggregating primitive types.

I chose snake_case here instead of camelCase because it seems to be used in the majority of the multi-word fns.

Do we also want to add these to `functions.py`?

This approach was recommended here: https://github.com/apache/spark/pull/8592#issuecomment-154247089

@marmbrus @rxin

### CODE REVIEW

**Comment by marmbrus** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 1]
```diff
@@ -175,6 +175,26 @@ object functions {
...
 
   /**
+   * Aggregate function: returns a list of objects with duplicates.
+   *
+   * For now this is an alias for the collect_list Hive UDAF.
+   *
+   * @group agg_funcs
+   * @since 1.6.0
+   */
+  def collect_list(e: Column): Column = callUdf("collect_list", e)
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: I think `callUdf` is deprecated and should be `callUDF`

**Comment by marmbrus** [comment]
Please also add them to functions.py and add unit tests for both languages.  Thanks!

Also, when you are done update the description as that becomes the commit message.

**Comment by nburoojy** [comment]
I've added these fns to pyspark and wrote unit tests. Which description needs to be updated?

**Comment by marmbrus** [comment]
Thanks, merging to master and 1.6

**Comment by marmbrus** [comment]
Hey, i just noticed this wasn't opened against master.  In the future please always go against master unless creating a specific backport.  We will cherry-pick when committing.

**Comment by maver1ck** [comment]
Hi,
How can I run this ?
Spark 1.6.0-preview1
Compiled with:
mvn -e -Pyarn -Phadoop-2.6 -Dhadoop.version=2.6.0 -DskipTests clean package

I'm trying like this:
    from pyspark.sql.functions import collect_list
    df.select(collect_list(df.client_id))

But I'm getting:
Py4JJavaError: An error occurred while calling o42.select.
: org.apache.spark.sql.AnalysisException: undefined function collect_list;
	at org.apache.spark.sql.catalyst.analysis.SimpleFunctionRegistry$$anonfun$2.apply(FunctionRegistry.scala:65)
	at org.apache.spark.sql.catalyst.analysis.SimpleFunctionRegistry$$anonfun$2.apply(FunctionRegistry.scala:65

**Comment by nburoojy** [comment]
This is a wrapper around the Hive collect fns. Try compiling with `-Phive -Phive-thriftserver`

**Comment by maver1ck** [comment]
Thanks.
And when I compile with hive is there a chance to do sth like this ?
"select id, collect_list(table.*) as data from table group by id" ?

**Pull request closed without merging**